File: OS_kernel/js/08_metabolism_resource.js
/**
 * ============================================================
 * KERNEL AGENTIX OS — 08_metabolism_resource.js
 * Module 8: Metabolism & Resource
 * Manajemen Baterai/Memori Berbasis Valensi
 * ============================================================
 * Depends on: corekernel.js
 *
 * Metabolism: agents consume and regenerate energy (battery).
 * Resource budget constrains action intensity.
 * No if-then: resource gates are continuous sigmoid masks.
 */
import { AgentixKernel, Semantic } from './corekernel.js';

class MetabolismResource extends AgentixKernel {
    constructor(config) {
        super(config);
        this._batteries  = {};   // { agentId: { charge, capacity, regen } }
        this._memory     = {};   // { agentId: { used, capacity } }
        this._budgets    = {};   // { agentId: { [actionType]: cost } }
    }

    // ─────────────────────────────────────────
    // BATTERY — energy store per agent
    // ─────────────────────────────────────────
    registerBattery(agentId, { capacity = 1.0, initial = 1.0, regen = 0.005 } = {}) {
        this._batteries[agentId] = {
            charge:   Math.min(initial, capacity),
            capacity,
            regen,
        };
    }

    getCharge(agentId) {
        return this._batteries[agentId]?.charge ?? 1.0;
    }

    // Charge as normalized valence [0,1]
    chargeValence(agentId) {
        const b = this._batteries[agentId];
        if (!b) return 1.0;
        return Semantic.normalize(b.charge, b.capacity);
    }

    // ─────────────────────────────────────────
    // CONSUME — drain battery (masked, no if)
    // Returns actual amount consumed
    // ─────────────────────────────────────────
    consume(agentId, amount) {
        const b = this._batteries[agentId];
        if (!b) return 0;
        const actual  = Math.min(amount, b.charge);
        b.charge      = Math.max(0, b.charge - actual);
        return actual;
    }

    // ─────────────────────────────────────────
    // REGENERATE — passive energy recovery each tick
    // ─────────────────────────────────────────
    regenerate(agentId) {
        const b = this._batteries[agentId];
        if (!b) return;
        b.charge = Math.min(b.capacity, b.charge + b.regen);
    }

    regenerateAll() {
        Object.keys(this._batteries).forEach(id => this.regenerate(id));
    }

    // ─────────────────────────────────────────
    // ENERGY GATE — soft mask on action intensity
    // Low charge → reduces output proportionally
    // threshold: below this charge, output fades
    // ─────────────────────────────────────────
    energyGate(agentId, threshold = 0.2, sharpness = 8) {
        const v = this.chargeValence(agentId);
        return Semantic.softThresh(v, threshold, sharpness);
    }

    // ─────────────────────────────────────────
    // ACTION BUDGET — cost registry per action type
    // ─────────────────────────────────────────
    registerBudget(agentId, costMap) {
        // costMap: { sprint: 0.02, jump: 0.05, idle: 0.001 }
        this._budgets[agentId] = { ...costMap };
    }

    getActionCost(agentId, actionType) {
        return this._budgets[agentId]?.[actionType] ?? 0;
    }

    // ─────────────────────────────────────────
    // WEIGHTED ACTION — apply cost and return
    // masked intensity (no if-then affordance check)
    // ─────────────────────────────────────────
    affordAction(agentId, actionType, desiredIntensity = 1.0) {
        const cost   = this.getActionCost(agentId, actionType);
        const gate   = this.energyGate(agentId);
        const actual = desiredIntensity * gate;
        this.consume(agentId, cost * actual);
        return actual;
    }

    // ─────────────────────────────────────────
    // MEMORY RESOURCE — token budget for agent memory
    // ─────────────────────────────────────────
    registerMemory(agentId, { capacity = 100 } = {}) {
        this._memory[agentId] = { used: 0, capacity };
    }

    allocMemory(agentId, tokens) {
        const m = this._memory[agentId];
        if (!m) return 0;
        const free    = m.capacity - m.used;
        const actual  = Math.min(tokens, free);
        m.used       += actual;
        return actual;
    }

    freeMemory(agentId, tokens) {
        const m = this._memory[agentId];
        if (!m) return;
        m.used = Math.max(0, m.used - tokens);
    }

    memoryPressure(agentId) {
        const m = this._memory[agentId];
        if (!m) return 0;
        return Semantic.normalize(m.used, m.capacity);
    }

    // ─────────────────────────────────────────
    // RESOURCE SUMMARY — snapshot of all agents
    // ─────────────────────────────────────────
    resourceSnapshot() {
        const out = {};
        Object.keys(this._batteries).forEach(id => {
            out[id] = {
                chargeValence: this.chargeValence(id),
                memoryPressure: this.memoryPressure(id),
            };
        });
        return out;
    }
}

export { MetabolismResource };

File: OS_kernel/py/08_metabolism_resource.py
"""
============================================================
KERNEL AGENTIX OS — 08_metabolism_resource.py
Module 8: Metabolism & Resource
Manajemen Baterai/Memori Berbasis Valensi
============================================================
Depends on: corekernel.py
"""
from corekernel import AgentixKernel, Semantic


class MetabolismResource(AgentixKernel):
    def __init__(self, config=None):
        super().__init__(config)
        self._batteries = {}
        self._memory    = {}
        self._budgets   = {}

    def register_battery(self, agent_id, capacity=1.0, initial=1.0, regen=0.005):
        self._batteries[agent_id] = {
            'charge':   min(initial, capacity),
            'capacity': capacity,
            'regen':    regen,
        }

    def get_charge(self, agent_id):
        return self._batteries.get(agent_id, {}).get('charge', 1.0)

    def charge_valence(self, agent_id):
        b = self._batteries.get(agent_id)
        if not b: return 1.0
        return Semantic.normalize(b['charge'], b['capacity'])

    def consume(self, agent_id, amount):
        b = self._batteries.get(agent_id)
        if not b: return 0.0
        actual     = min(amount, b['charge'])
        b['charge'] = max(0.0, b['charge'] - actual)
        return actual

    def regenerate(self, agent_id):
        b = self._batteries.get(agent_id)
        if b: b['charge'] = min(b['capacity'], b['charge'] + b['regen'])

    def regenerate_all(self):
        for aid in self._batteries: self.regenerate(aid)

    def energy_gate(self, agent_id, threshold=0.2, sharpness=8):
        return Semantic.soft_thresh(self.charge_valence(agent_id), threshold, sharpness)

    def register_budget(self, agent_id, cost_map):
        """cost_map: { 'sprint': 0.02, 'jump': 0.05, 'idle': 0.001 }"""
        self._budgets[agent_id] = dict(cost_map)

    def get_action_cost(self, agent_id, action_type):
        return self._budgets.get(agent_id, {}).get(action_type, 0.0)

    def afford_action(self, agent_id, action_type, desired_intensity=1.0):
        cost   = self.get_action_cost(agent_id, action_type)
        gate   = self.energy_gate(agent_id)
        actual = desired_intensity * gate
        self.consume(agent_id, cost * actual)
        return actual

    def register_memory(self, agent_id, capacity=100):
        self._memory[agent_id] = {'used': 0, 'capacity': capacity}

    def alloc_memory(self, agent_id, tokens):
        m = self._memory.get(agent_id)
        if not m: return 0
        free   = m['capacity'] - m['used']
        actual = min(tokens, free)
        m['used'] += actual
        return actual

    def free_memory(self, agent_id, tokens):
        m = self._memory.get(agent_id)
        if m: m['used'] = max(0, m['used'] - tokens)

    def memory_pressure(self, agent_id):
        m = self._memory.get(agent_id)
        if not m: return 0.0
        return Semantic.normalize(m['used'], m['capacity'])

    def resource_snapshot(self):
        return {
            aid: {
                'charge_valence':  self.charge_valence(aid),
                'memory_pressure': self.memory_pressure(aid),
            }
            for aid in self._batteries
        }
